<?php

declare(strict_types=1);

namespace Atk4\Ui\Tests;

use Atk4\Core\Phpunit\TestCase;
use Atk4\Data\Model;
use Atk4\Ui\App;
use Atk4\Ui\Form;

class FormTest extends TestCase
{
    /** @var Form */
    public $f;

    /** @var string */
    public $f_error;

    protected function setUp(): void
    {
        parent::setUp(); // TODO: Change the autogenerated stub

        $this->f = new \Atk4\Ui\Form();
        $this->f->setApp(new AppFormTestMock([
            'catch_exceptions' => false,
            'always_run' => false,
            'catch_runaway_callbacks' => false,
        ]));
        $this->f->invokeInit();
    }

    /**
     * Some tests for form.
     */
    public function testGetField(): void
    {
        $f = $this->f;
        $f->addControl('test');

        $this->assertTrue($f->getControl('test') instanceof Form\Control);
        $this->assertInstanceOf(Form\Control::class, $f->layout->getControl('test'));
    }

    public function assertSubmit(array $post_data, \Closure $submit = null, \Closure $check_expected_error = null)
    {
        $submit_called = false;
        $_POST = $post_data;
        // trigger callback
        $_GET['atk_submit'] = 'ajax';
        $_GET['__atk_callback'] = 'atk_submit';

        $this->f->onSubmit(function (Form $form) use (&$submit_called, $submit) {
            $submit_called = true;
            if ($submit) {
                $submit($form->model);
            }
        });

        $this->f->render();
        $res = $this->f->getApp()->output;

        if ($check_expected_error) {
            $this->assertFalse($submit_called, 'Expected submission to fail, but it was successful!');
            $this->assertNotSame('', $res['atkjs']); // will output useful error
            $this->f_error = $res['atkjs'];

            $check_expected_error($res['atkjs']);
        } else {
            $this->assertTrue($submit_called, 'Expected submission to be successful but it failed');
            $this->assertSame('', $res['atkjs']); // will output useful error
        }

        $this->f = null;   // we shouldn't submit form twice!

        $_POST = [];
    }

    public function testFormSubmit(): void
    {
        $f = $this->f;

        $m = new Model();
        $m->addField('name', ['default' => 'John']);
        $m->addField('email', ['required' => true]);
        $m->addField('is_admin', ['default' => false]);

        $m = $m->createEntity();
        $f->setModel($m, ['name', 'email']);

        $this->assertSame('John', $f->model->get('name'));

        // fake some POST data
        $this->assertSubmit(['email' => 'john@yahoo.com', 'is_admin' => '1'], function (Model $m) {
            // field has default, but form didn't send value back
            $this->assertNull($m->get('name'));

            $this->assertSame('john@yahoo.com', $m->get('email'));

            // security check, unspecified field must not be changed
            $this->assertFalse($m->get('is_admin'));
        });
    }

    public function testTextarea(): void
    {
        $this->f->addControl('Textarea');
        $this->assertSubmit(['Textarea' => '0'], function (Model $m) {
            $this->assertSame('0', $m->get('Textarea'));
        });
    }

    public function assertSubmitError(array $post, \Closure $error_callback)
    {
        $this->assertSubmit($post, null, $error_callback);
    }

    public function assertFormControlError(string $field, string $error)
    {
        $matched = false;

        preg_replace_callback('/form\("add prompt","([^"]*)","([^"]*)"\)/', function ($matches) use ($error, $field, &$matched) {
            if ($matches[1] === $field) {
                $this->assertStringContainsString($error, $matches[2], 'Regarding control ' . $field . ' error message');

                $matched = true;
            }
        }, $this->f_error);

        $this->assertTrue($matched, 'Form control ' . $field . ' did not produce error');
    }

    public function assertFromControlNoErrors(string $field)
    {
        preg_replace_callback('/form\("add prompt","([^"]*)","([^"]*)"\)/', function ($matches) use ($field) {
            if ($matches[1] === $field) {
                $this->fail('Form control ' . $field . ' unexpected error: ' . $matches[2]);
            }
        }, $this->f_error);
    }

    public function testSubmitError(): void
    {
        $m = new Model();

        $options = ['0' => 'yes please', '1' => 'woot'];

        $m->addField('opt1', ['values' => $options]);
        $m->addField('opt2', ['values' => $options]);
        $m->addField('opt3', ['values' => $options, 'required' => true]);
        //$m->addField('opt3_zerotest', ['values'=>$options, 'required'=>true]);
        $m->addField('opt4', ['values' => $options, 'mandatory' => true]);

        $m = $m->createEntity();
        $this->f->setModel($m);
        $this->assertSubmitError(['opt1' => '2', 'opt3' => '', 'opt3_zerotest' => '0'], function ($error) {
            // dropdown validates to make sure option is proper
            $this->assertFormControlError('opt1', 'not one of the allowed values');

            // user didn't select any option here
            $this->assertFromControlNoErrors('opt2');

            // dropdown insists for value to be there
            $this->assertFormControlError('opt3', 'Must not be empty');

            // value with '0' is valid selection
            // TODO: currently fails!! See https://github.com/atk4/ui/issues/781
            //$this->assertFromControlNoErrors('opt3_zerotest');

            // mandatory will error during save(), but form does not care about it. This is normal
            // as there may be further changes to this field on beforeSave hook...
            $this->assertFromControlNoErrors('opt4');
        });
    }
}

class AppFormTestMock extends App
{
    public $output;

    public function terminate($output = '', array $headers = []): void
    {
        $this->output = $output;
    }
}
